client/src/lib/attempt-result.js

```diff
@@ -124,6 +139,107 @@ function mean(values) {
   return Math.round(sum / values.length);
 }
 
+/**
+ * return the number of competitors that should be marked as advancing.
+ * When no competitors advance, return 3 for the podium placements.
+ * Might not be accurate for percentage based advancement but that is acceptable
+ */
+function getAdvancingCount(results, advancementCondition) {
+  if (!advancementCondition) return 3;
+  if (advancementCondition.type === "ranking") {
+    return advancementCondition.level;
+  }
+  return results.filter((result) => result.advancing || result.advancingQuestionable).length;
+}
+
+/**
+ * return projectedAverage based on current results.
+ * projections are defined as follows:
+ *   - MO3 events: mean of current solves
+ *   - AVG5 events:
+ *     - 1-2 solves: mean of current solves
+ *     - 3-4 solves: median of current solves
+ */
+export function computeProjectedAverage(result, format) {
+  const attemptResults = result.attempts.map((attempt) => attempt.result);
+  if (result.average) {
+    return result.average;
+  }
+  if (attemptResults.length > 0) {
+    if (format.numberOfAttempts === 3 || attemptResults.length < 3) {
+      return meanOfX(attemptResults);
+    }
+    if (attemptResults.length === 3) {
+      const [, x,] = attemptResults.slice().sort(compareAttemptResults);
+      return x;
+    }
+    const [, x, y,] = attemptResults.slice().sort(compareAttemptResults);
+    return meanOfX([x, y]);
+  }
+  return SKIPPED_VALUE;
+}
+
+/**
+ * return the results object with additional properties:
+ * projectedAverage: projected final average based on current results
+ * forFirst: time needed to overtake first place
+ * forThird: time needed to overtake third place
+ */
+export function getExpandedResults(results, format, forecastView, advancementCondition) {
+  if (results.length == 0 || !forecastView) return results;
+  var expandedResults = results.map((result) => {
+    return {
+      ...result,
+      projectedAverage: computeProjectedAverage(result, format),
+      forFirst: SKIPPED_VALUE,
+      forThird: SKIPPED_VALUE,
+    };
+  });
+  
+  if (expandedResults[0].projectedAverage == SKIPPED_VALUE) {
+    // First place has no results. Do nothing.
+    return expandedResults;
+  }
+
+  for (let result of expandedResults) {
+    result.advancing = false;
+    result.advancingQuestionable = false;
+  }
+
+  const advancingCount = getAdvancingCount(results, advancementCondition);
+  const roundIncomplete = results.some((result) => isSkipped(result.average));
+
+  // Sort based on projection with tiebreakers on single
+  expandedResults = expandedResults.sort((a, b) => compareProjectedResults(a, b));
```

Comment:
We can simplify a bit, first define this for attempt result:

```js
function toMonotonic(attemptResult) {
  if (isComplete(attemptResult)) {
    return attemptResult;
  } else {
    return Infinity;
  }
}
```

and then here we can do:

```js
expandedResults =
  orderBy(expandedResults, [
    (result) => toMonotonic(result.projectedAverage),
    (result) => toMonotonic(result.best),
  ]);
```